tests/test_agents/base.py

```diff
@@ -465,22 +465,27 @@ def _stop_and_restart_last_agent(self, nb_agents: int) -> None:
             timeout=self.wait_before_stop,
         )
         if missing_strict_strings:
-            raise RuntimeError("cannot stop agent correctly")
-        logging.info("Last agent stopped")
+            msg = f"cannot stop agent, stop string `{self.stop_string}` not found"
+            raise RuntimeError(msg)
+
+        if self.wait_to_kill:
+            logging.info("Waiting to terminate agents")
+            time.sleep(self.wait_to_kill)
 
-        # immediately terminate the agent
-        self.terminate_agents(self.processes[-1], timeout=0)
+        # immediately terminate the agents - sequentially
         # don't pop before termination, seems to lead to failure!
-        self.processes.pop()
+        for _ in range(self.n_terminal):
+            self.terminate_agents(self.processes[-1], timeout=0)
+            self.processes.pop()
+            logging.info("Last agent stopped")
 
         # wait for some time before restarting
         logging.info(
             f"Waiting {self.restart_after} seconds before restarting the agent"
         )
         time.sleep(self.restart_after)
 
-        # restart agent
-        logging.info("Restart the agent")
-        self._launch_agent_i(-1, nb_agents)
-        self.check_aea_messages()
-        self.terminate_processes()
+        # restart agents
+        for i in range(self.n_terminal):
+            logging.info(f"Restart the agent {~i}")
+            self._launch_agent_i(~i, self.nb_nodes)
```

Comment:
Why not `-i`?

Reply:
not exactly the same, would need to `i - 1`. Best idea, though, is to not loop backwards in the first place, as you suggested we don't need to start at the end, or use reverse(range(1, len(x) + 1)) or so (which is also not nice).